// @visactor/vchart-extension/cjs/components/histogram-regression-line/index.js
"use strict";

var __rest = this && this.__rest || function(s, e) {
    var t = {};
    for (var p in s) Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0 && (t[p] = s[p]);
    if (null != s && "function" == typeof Object.getOwnPropertySymbols) {
        var i = 0;
        for (p = Object.getOwnPropertySymbols(s); i < p.length; i++) e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]) && (t[p[i]] = s[p[i]]);
    }
    return t;
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.appendHistogramRegressionLineConfig = exports.getHistogramRegressionLineConfig = void 0;

const vchart_1 = require("@visactor/vchart"), vchart_2 = require("@visactor/vchart"), regression_line_1 = require("../regression-line/regression-line"), getRegressionByType = (type, data, kdeOptions) => {
    switch (type) {
      case "kde":
        return (0, vchart_1.kde)(data, kdeOptions);

      case "ecdf":
        return (0, vchart_1.ecdf)(data);
    }
};

function getHistogramRegressionLineConfig(type, config) {
    const {line: line, label: label, color: color} = config;
    return {
        type: "component",
        componentType: regression_line_1.REGRESSION_LINE,
        interactive: !1,
        zIndex: 500,
        style: {
            data: (datum, ctx) => {
                const series = ctx.vchart.getChart().getAllSeries().filter((s => s.type === vchart_2.SeriesTypeEnum.bar)), regressionData = [];
                return series && series.length && series.forEach((s => {
                    var _a, _b, _c, _d, _e, _f, _g, _h;
                    const region = s.getRegion().getLayoutStartPoint(), rawData = s._rawData, data = null == rawData ? void 0 : rawData.rawData, binTransformOptions = null === (_b = null === (_a = rawData.transformsArr) || void 0 === _a ? void 0 : _a.find((t => "bin" === t.type))) || void 0 === _b ? void 0 : _b.options, fieldX = null === (_c = s.fieldX) || void 0 === _c ? void 0 : _c[0], scaleY = s.getYAxisHelper().getScale(0), viewData = s.getViewData().latestData;
                    if (!(data && data.length && (null == binTransformOptions ? void 0 : binTransformOptions.field) && scaleY && viewData && viewData.length)) return;
                    const simpleData = data.map((entry => entry[binTransformOptions.field])), res = getRegressionByType(type, simpleData, "kde" === type ? {
                        bandwidth: viewData[0][null !== (_e = null === (_d = binTransformOptions.outputNames) || void 0 === _d ? void 0 : _d.x1) && void 0 !== _e ? _e : "x1"] - viewData[0][null !== (_g = null === (_f = binTransformOptions.outputNames) || void 0 === _f ? void 0 : _f.x0) && void 0 !== _g ? _g : "x0"]
                    } : null), N = Math.max(3, Math.floor(simpleData.length / 4)), lineData = res.evaluateGrid(N), yRange = scaleY.range(), y0 = yRange[0], y1 = (0, 
                    vchart_1.last)(yRange), scaleR = "kde" === type ? k => scaleY.scale(k * data.length * res.bandwidth) : e => y0 + (y1 - y0) * e;
                    regressionData.push({
                        color: null != color ? color : null === (_h = s.getOption().globalScale.getScale("color")) || void 0 === _h ? void 0 : _h.scale(s.getSeriesKeys()[0]),
                        line: lineData.map((ld => {
                            const d = {
                                [fieldX]: ld.x
                            };
                            return {
                                x: s.dataToPositionX(d) + region.x,
                                y: scaleR(ld.y) + region.y
                            };
                        }))
                    });
                })), regressionData;
            },
            line: line,
            label: label
        }
    };
}

function appendHistogramRegressionLineConfig(chartSpec, spec) {
    var _a, _b;
    spec || (spec = null !== (_a = (0, vchart_1.get)(chartSpec, regression_line_1.REGRESSION_LINE)) && void 0 !== _a ? _a : (0, 
    vchart_1.get)(null === (_b = chartSpec.series) || void 0 === _b ? void 0 : _b.find((s => s.type === vchart_2.SeriesTypeEnum.bar)), regression_line_1.REGRESSION_LINE));
    (0, vchart_1.array)(spec).forEach((s => {
        if (!1 !== s.visible) {
            chartSpec.customMark || (chartSpec.customMark = []);
            const {type: type} = s, rest = __rest(s, [ "type" ]);
            chartSpec.customMark.push(getHistogramRegressionLineConfig(type, rest));
        }
    }));
}

exports.getHistogramRegressionLineConfig = getHistogramRegressionLineConfig, exports.appendHistogramRegressionLineConfig = appendHistogramRegressionLineConfig;
