// @visactor/vchart-extension/cjs/components/scatter-regression-line/index.js
"use strict";

var __rest = this && this.__rest || function(s, e) {
    var t = {};
    for (var p in s) Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0 && (t[p] = s[p]);
    if (null != s && "function" == typeof Object.getOwnPropertySymbols) {
        var i = 0;
        for (p = Object.getOwnPropertySymbols(s); i < p.length; i++) e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]) && (t[p[i]] = s[p[i]]);
    }
    return t;
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.appendScatterRegressionLineConfig = exports.getScatterRegressionLineConfig = void 0;

const vchart_1 = require("@visactor/vchart"), vchart_2 = require("@visactor/vchart"), regression_line_1 = require("../regression-line/regression-line"), getRegressionByType = (type, data, x = (d => d.x), y = (d => d.y), degree) => {
    switch (type) {
      case "logisitc":
        return (0, vchart_1.regressionLogistic)(data, x, y);

      case "lowess":
        return (0, vchart_1.regressionLowess)(data, x, y);

      case "polynomial":
        return (0, vchart_1.regressionPolynomial)(data, x, y, {
            degree: degree
        });

      default:
        return (0, vchart_1.regressionLinear)(data, x, y);
    }
};

function getScatterRegressionLineConfig(type, config) {
    const {color: color, line: line, confidenceInterval: confidenceInterval, label: label} = config;
    return {
        type: "component",
        componentType: regression_line_1.REGRESSION_LINE,
        interactive: !1,
        style: {
            data: (datum, ctx) => {
                const series = ctx.vchart.getChart().getAllSeries().filter((s => s.type === vchart_2.SeriesTypeEnum.scatter)), regressionData = [];
                return series && series.length && series.forEach((s => {
                    var _a, _b;
                    const start = s.getRegion().getLayoutStartPoint(), rect = s.getRegion().getLayoutRect(), yClamper = (0, 
                    vchart_1.clamper)(start.y, start.y + rect.height), colorAttrOptions = s.getColorAttribute(), groups = s.getSeriesKeys(), data = s.getViewData().latestData, fieldX = null === (_a = s.fieldX) || void 0 === _a ? void 0 : _a[0], fieldY = null === (_b = s.fieldY) || void 0 === _b ? void 0 : _b[0];
                    fieldX && fieldY && data && !(data.length <= 2) && groups.forEach((group => {
                        var _a;
                        const groupData = data.filter((d => d[null == colorAttrOptions ? void 0 : colorAttrOptions.field] === group));
                        if (!groupData.length) return;
                        const {evaluateGrid: evaluateGrid, confidenceInterval: confidenceInterval} = getRegressionByType(type, groupData, (datum => null == datum ? void 0 : datum[fieldX]), (datum => null == datum ? void 0 : datum[fieldY]), config.polynomialDegree), N = Math.max(3, Math.floor(groupData.length / 4)), lineData = evaluateGrid(N), confidenceData = confidenceInterval(N);
                        regressionData.push({
                            color: null != color ? color : null === (_a = null == colorAttrOptions ? void 0 : colorAttrOptions.scale) || void 0 === _a ? void 0 : _a.scale(group),
                            line: lineData.map((ld => {
                                const d = {
                                    [fieldX]: ld.x,
                                    [fieldY]: ld.y
                                };
                                return {
                                    x: s.dataToPositionX(d) + start.x,
                                    y: yClamper("logisitc" === type ? start.y + rect.height * (1 - ld.y) : s.dataToPositionY(d) + start.y)
                                };
                            })),
                            area: confidenceData.map((c => {
                                const d = {
                                    [fieldX]: c.x,
                                    [fieldY]: c.lower
                                };
                                return {
                                    x: s.dataToPositionX(d) + start.x,
                                    y: yClamper(s.dataToPositionY(d) + start.y),
                                    y1: yClamper(s.dataToPositionY({
                                        [fieldY]: c.upper
                                    }) + start.y)
                                };
                            }))
                        });
                    }));
                })), regressionData;
            },
            line: line,
            confidenceInterval: confidenceInterval,
            label: label
        }
    };
}

function appendScatterRegressionLineConfig(chartSpec, spec) {
    var _a, _b;
    spec || (spec = null !== (_a = (0, vchart_1.get)(chartSpec, regression_line_1.REGRESSION_LINE)) && void 0 !== _a ? _a : (0, 
    vchart_1.get)(null === (_b = chartSpec.series) || void 0 === _b ? void 0 : _b.find((s => s.type === vchart_2.SeriesTypeEnum.scatter)), regression_line_1.REGRESSION_LINE));
    (0, vchart_1.array)(spec).forEach((s => {
        if (!1 !== s.visible) {
            chartSpec.customMark || (chartSpec.customMark = []);
            const {type: type} = s, rest = __rest(s, [ "type" ]);
            chartSpec.customMark.push(getScatterRegressionLineConfig(type, rest));
        }
    }));
}

exports.getScatterRegressionLineConfig = getScatterRegressionLineConfig, exports.appendScatterRegressionLineConfig = appendScatterRegressionLineConfig;
