// @visactor/vchart-extension/esm/components/regression-line/regression-line.js
import { AbstractComponent, createText, createLine, createArea } from "@visactor/vchart";

import { isEmpty, isValid, last } from "@visactor/vchart";

import { Factory } from "@visactor/vchart";

export const REGRESSION_LINE = "regressionLine";

export class RegressionLine extends AbstractComponent {
    constructor() {
        super(...arguments), this.name = REGRESSION_LINE;
    }
    render() {
        this.removeAllChild();
        const {data: data, line: line = {}, label: label, name: name = "regression-line", confidenceInterval: confidenceInterval} = this.attribute;
        isEmpty(data) || data.forEach((d => {
            const {color: color} = d;
            if (d.area && !1 !== (null == confidenceInterval ? void 0 : confidenceInterval.visible)) {
                const areaShape = createArea(Object.assign(Object.assign({
                    points: d.area
                }, isValid(color) ? {
                    fill: color,
                    fillOpacity: .12
                } : null), null == confidenceInterval ? void 0 : confidenceInterval.style));
                areaShape.name = "scatter-regression-area", this.add(areaShape);
            }
            if (d.line && !1 !== (null == line ? void 0 : line.visible)) {
                const lineShape = createLine(Object.assign(Object.assign({
                    points: d.line,
                    lineWidth: 1
                }, isValid(color) ? {
                    stroke: color
                } : null), null == line ? void 0 : line.style));
                lineShape.name = `${name}-curve`, this.add(lineShape);
            }
            const lastPoint = last(d.line);
            if (label && !1 !== label.visible && label.text && lastPoint) {
                const tag = createText(Object.assign(Object.assign(Object.assign({}, lastPoint), {
                    text: label.text,
                    textAlign: "end",
                    textBaseline: "middle"
                }), label.style));
                tag.name = `${name}-label`, this.add(tag);
            }
        }));
    }
}

export const registerRegressionLine = () => {
    Factory.registerGraphicComponent(REGRESSION_LINE, (attrs => new RegressionLine(attrs)));
};
