// @visactor/vchart-extension/cjs/charts/3d/rect-3d.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerRect3dMark = exports.Rect3dMark = void 0;

const vchart_1 = require("@visactor/vchart"), vchart_2 = require("@visactor/vchart");

class Rect3dMark extends vchart_2.BaseMark {
    _getDefaultStyle() {
        return Object.assign(Object.assign({}, super._getDefaultStyle()), {
            width: void 0,
            height: void 0,
            length: 3
        });
    }
    setDataLabelType() {
        return "rect";
    }
    constructor(name, option) {
        super(name, option), this.type = Rect3dMark.type, this._markConfig.support3d = !0;
    }
    setMarkConfig(config) {
        super.setMarkConfig(config), this._markConfig.support3d = !0;
    }
}

exports.Rect3dMark = Rect3dMark, Rect3dMark.type = "rect3d";

const registerRect3dMark = () => {
    vchart_2.Factory.registerMark(Rect3dMark.type, Rect3dMark), (0, vchart_1.registerShadowRoot)(), 
    (0, vchart_1.registerRect3d)(), (0, vchart_2.registerRectAnimation)(), vchart_2.Factory.registerGraphicComponent("rect3d", vchart_1.createRect3d);
};

exports.registerRect3dMark = registerRect3dMark;
