// @visactor/vchart-extension/esm/components/bar-regression-line/index.js
var __rest = this && this.__rest || function(s, e) {
    var t = {};
    for (var p in s) Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0 && (t[p] = s[p]);
    if (null != s && "function" == typeof Object.getOwnPropertySymbols) {
        var i = 0;
        for (p = Object.getOwnPropertySymbols(s); i < p.length; i++) e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]) && (t[p[i]] = s[p[i]]);
    }
    return t;
};

import { array, clamper, get, regressionPolynomial } from "@visactor/vchart";

import { SeriesTypeEnum } from "@visactor/vchart";

import { REGRESSION_LINE } from "../regression-line/regression-line";

export function getBarRegressionLineConfig(config) {
    const {color: color, line: line, confidenceInterval: confidenceInterval, label: label} = config;
    return {
        type: "component",
        componentType: REGRESSION_LINE,
        interactive: !1,
        zIndex: 500,
        style: {
            data: (datum, ctx) => {
                const series = ctx.vchart.getChart().getAllSeries().filter((s => s.type === SeriesTypeEnum.bar)), regressionData = [];
                return series && series.length && series.forEach((s => {
                    var _a, _b, _c;
                    const region = s.getRegion().getLayoutStartPoint(), start = s.getRegion().getLayoutStartPoint(), rect = s.getRegion().getLayoutRect(), yClamper = clamper(start.y, start.y + rect.height), data = s.getViewData().latestData, fieldX = null === (_a = s.fieldX) || void 0 === _a ? void 0 : _a[0], fieldY = null === (_b = s.fieldY) || void 0 === _b ? void 0 : _b[0], isHorizontal = "horizontal" === s.direction, groups = s.getRawDataStatisticsByField(fieldX).values;
                    if (isHorizontal || !fieldX || !fieldY || !data || data.length <= 2 || groups.length <= 2) return;
                    const {evaluateGrid: evaluateGrid, confidenceInterval: confidenceInterval} = regressionPolynomial(data, (datum => groups.indexOf(null == datum ? void 0 : datum[fieldX])), (datum => null == datum ? void 0 : datum[fieldY]), {
                        degree: config.degree
                    }), N = groups.length, lineData = evaluateGrid(N), confidenceData = confidenceInterval(N), halfBandWidth = s.getXAxisHelper().getBandwidth(0) / 2;
                    regressionData.push({
                        color: null != color ? color : null === (_c = s.getOption().globalScale.getScale("color")) || void 0 === _c ? void 0 : _c.scale(s.getSeriesKeys()[0]),
                        line: lineData.map((ld => {
                            const d = {
                                [fieldX]: groups[ld.x],
                                [fieldY]: ld.y
                            };
                            return {
                                x: s.dataToPositionX(d) + region.x + halfBandWidth,
                                y: yClamper(s.dataToPositionY(d) + region.y)
                            };
                        })),
                        area: confidenceData.map((c => {
                            const d = {
                                [fieldX]: groups[c.x],
                                [fieldY]: c.lower
                            };
                            return {
                                x: s.dataToPositionX(d) + region.x + halfBandWidth,
                                y: yClamper(s.dataToPositionY(d) + region.y),
                                y1: yClamper(s.dataToPositionY({
                                    [fieldY]: c.upper
                                }) + region.y)
                            };
                        }))
                    });
                })), regressionData;
            },
            line: line,
            confidenceInterval: confidenceInterval,
            label: label
        }
    };
}

export function appendBarRegressionLineConfig(chartSpec, spec) {
    var _a, _b;
    spec || (spec = null !== (_a = get(chartSpec, REGRESSION_LINE)) && void 0 !== _a ? _a : get(null === (_b = chartSpec.series) || void 0 === _b ? void 0 : _b.find((s => s.type === SeriesTypeEnum.bar)), REGRESSION_LINE));
    array(spec).forEach((s => {
        if (!1 !== s.visible) {
            chartSpec.customMark || (chartSpec.customMark = []);
            const {visible: visible} = s, rest = __rest(s, [ "visible" ]);
            chartSpec.customMark.push(getBarRegressionLineConfig(rest));
        }
    }));
}
