// @code-inspector/webpack/dist/index.mjs
var L = Object.defineProperty, N = Object.defineProperties;
var O = Object.getOwnPropertyDescriptors;
var C = Object.getOwnPropertySymbols;
var S = Object.prototype.hasOwnProperty, z = Object.prototype.propertyIsEnumerable;
var W = (s, t, e) => t in s ? L(s, t, { enumerable: !0, configurable: !0, writable: !0, value: e }) : s[t] = e, d = (s, t) => {
  for (var e in t || (t = {}))
    S.call(t, e) && W(s, e, t[e]);
  if (C)
    for (var e of C(t))
      z.call(t, e) && W(s, e, t[e]);
  return s;
}, v = (s, t) => N(s, O(t));
var c = (s, t, e) => new Promise((n, a) => {
  var r = (o) => {
    try {
      i(e.next(o));
    } catch (f) {
      a(f);
    }
  }, u = (o) => {
    try {
      i(e.throw(o));
    } catch (f) {
      a(f);
    }
  }, i = (o) => o.done ? n(o.value) : Promise.resolve(o.value).then(r, u);
  i((e = e.apply(s, t)).next());
});
import { normalizePath as _, fileURLToPath as I, isDev as R, getProjectRecord as F, isNextjsProject as H, getCodeWithWebComponent as Q } from "@code-inspector/core";
import j, { dirname as U } from "path";
function V(s, t) {
  return c(this, null, function* () {
    if (!s || !t)
      return [];
    const e = typeof s == "function" ? yield s() : s;
    let n = [];
    if (typeof e == "object" && !Array.isArray(e))
      for (const a in e) {
        const r = e[a], u = r.import || r;
        $(n, u, t);
      }
    else
      $(n, e, t);
    return n.filter((a) => !!a);
  });
}
function $(s, t, e) {
  typeof t == "string" ? s.push(w(t, e)) : Array.isArray(t) && s.push(
    ...t.map((n) => w(n, e))
  );
}
function w(s, t) {
  return j.isAbsolute(s) ? _(s) : s.startsWith(".") ? j.resolve(t, _(s)) : "";
}
let p = "";
typeof __dirname != "undefined" ? p = __dirname : p = U(I(import.meta.url));
let P = !0;
const q = (s, t) => {
  var r, u;
  if (!P)
    return;
  P = !1;
  const e = (t == null ? void 0 : t.compiler) || t, n = (r = e == null ? void 0 : e.options) == null ? void 0 : r.module;
  ((n == null ? void 0 : n.rules) || (n == null ? void 0 : n.loaders) || []).push(
    d({
      test: (u = s.match) != null ? u : /\.html$/,
      resourceQuery: /vue/,
      use: [
        {
          loader: j.resolve(p, "./loader.js"),
          options: s
        }
      ]
    }, s.enforcePre === !1 ? {} : { enforce: "pre" }),
    d({
      test: /\.(vue|jsx|tsx|js|ts|mjs|mts|svelte)$/,
      use: [
        {
          loader: j.resolve(p, "./loader.js"),
          options: s
        }
      ]
    }, s.enforcePre === !1 ? {} : { enforce: "pre" }),
    v(d({}, s.injectTo ? { resource: s.injectTo } : {
      test: /\.(jsx|tsx|js|ts|mjs|mts)$/,
      exclude: /node_modules/
    }), {
      use: [
        {
          loader: j.resolve(p, "./inject-loader.js"),
          options: s
        }
      ],
      enforce: H() ? "pre" : "post"
    })
  );
};
function D(s, t, e) {
  return Q({
    options: v(d({}, s), { importClient: "code" }),
    file: "main.js",
    code: "",
    record: t,
    inject: !0,
    server: e
  });
}
function B(n) {
  return c(this, arguments, function* ({
    options: s,
    record: t,
    assets: e
  }) {
    const a = Object.keys(e).filter((r) => /\.html$/.test(r));
    if (a.length) {
      const r = yield D(s, t);
      a.forEach((u) => {
        var o, f;
        const i = (f = (o = e[u]) == null ? void 0 : o.source) == null ? void 0 : f.call(o);
        if (typeof i == "string") {
          const h = i.replace(
            "<head>",
            '<head><script type="module">\n'.concat(r, "\n<\/script>")
          );
          e[u] = {
            source: () => h,
            size: () => h.length
          };
        }
      });
    }
  });
}
class M {
  constructor(t) {
    this.options = t;
  }
  apply(t) {
    return c(this, null, function* () {
      var a, r, u, i, o, f, h, b, k, A, y, x;
      if (P = !0, this.options.close || !R(
        this.options.dev,
        ((a = t == null ? void 0 : t.options) == null ? void 0 : a.mode) === "development" || process.env.NODE_ENV === "development"
      ))
        return;
      const e = {
        port: 0,
        entry: "",
        output: this.options.output,
        inputs: V(
          (r = t == null ? void 0 : t.options) == null ? void 0 : r.entry,
          (u = t == null ? void 0 : t.options) == null ? void 0 : u.context
        ),
        envDir: (i = t == null ? void 0 : t.options) == null ? void 0 : i.context,
        root: (o = t == null ? void 0 : t.options) == null ? void 0 : o.context
      }, n = ((f = t == null ? void 0 : t.options) == null ? void 0 : f.cache) || ((b = (h = t == null ? void 0 : t.options) == null ? void 0 : h.experiments) == null ? void 0 : b.cache);
      if ((n == null ? void 0 : n.type) === "filesystem" && (this.options.cache ? (e.port = this.options.port || ((k = F(e)) == null ? void 0 : k.previousPort), D(this.options, e, !0)) : n.version = "code-inspector-".concat(Date.now())), q(v(d({}, this.options), { record: e }), t), (A = t == null ? void 0 : t.hooks) != null && A.emit && !((x = (y = this.options.skipSnippets) == null ? void 0 : y.includes) != null && x.call(y, "htmlScript"))) {
        const E = this.options;
        t.hooks.emit.tapAsync(
          "WebpackCodeInspectorPlugin",
          (g, T) => c(this, null, function* () {
            let l = {};
            g.getAssets ? l = yield g.getAssets() : l = g.assets, yield B({
              options: E,
              record: e,
              assets: l
            }), T();
          })
        );
      }
    });
  }
}
export {
  M as default
};
