// @visactor/vrender-core/cjs/render/contributions/render/image-render.js
"use strict";

var __decorate = this && this.__decorate || function(decorators, target, key, desc) {
    var d, c = arguments.length, r = c < 3 ? target : null === desc ? desc = Object.getOwnPropertyDescriptor(target, key) : desc;
    if ("object" == typeof Reflect && "function" == typeof Reflect.decorate) r = Reflect.decorate(decorators, target, key, desc); else for (var i = decorators.length - 1; i >= 0; i--) (d = decorators[i]) && (r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r);
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}, __metadata = this && this.__metadata || function(k, v) {
    if ("object" == typeof Reflect && "function" == typeof Reflect.metadata) return Reflect.metadata(k, v);
}, __param = this && this.__param || function(paramIndex, decorator) {
    return function(target, key) {
        decorator(target, key, paramIndex);
    };
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.DefaultCanvasImageRender = exports.drawImageWithLayout = exports.shouldClipImageByLayout = exports.resolveImageRepeatMode = exports.resolveBackgroundParamsByImageSizing = exports.resolveImageMode = void 0;

const inversify_lite_1 = require("../../../common/inversify-lite"), contribution_provider_1 = require("../../../common/contribution-provider"), theme_1 = require("../../../graphic/theme"), constants_1 = require("./contributions/constants"), constants_2 = require("../../../graphic/constants"), vutils_1 = require("@visactor/vutils"), rect_1 = require("../../../common/shape/rect"), base_render_1 = require("./base-render"), contributions_1 = require("./contributions"), loader_1 = require("../../../resource-loader/loader"), base_contribution_render_1 = require("./contributions/base-contribution-render"), repeatStr = [ "", "repeat-x", "repeat-y", "repeat" ];

function resolveImageMode({repeatX: repeatX = "no-repeat", repeatY: repeatY = "no-repeat", imageMode: imageMode}) {
    const repeatMode = resolveImageRepeatMode(repeatX, repeatY);
    return {
        repeatMode: repeatMode,
        sizingMode: "no-repeat" === repeatMode && null != imageMode ? imageMode : "fill"
    };
}

exports.resolveImageMode = resolveImageMode;

const IMAGE_MODE_TO_BACKGROUND_MODE = {
    cover: "no-repeat-cover",
    contain: "no-repeat-contain",
    fill: "no-repeat-fill",
    auto: "no-repeat-auto"
};

function resolveBackgroundParamsByImageSizing(sizingMode) {
    return {
        backgroundMode: IMAGE_MODE_TO_BACKGROUND_MODE[sizingMode],
        backgroundFit: !1,
        backgroundKeepAspectRatio: !1
    };
}

function resolveImageRepeatMode(repeatX, repeatY) {
    let repeat = 0;
    return "repeat" === repeatX && (repeat |= 1), "repeat" === repeatY && (repeat |= 2), 
    repeat ? repeatStr[repeat] : "no-repeat";
}

function shouldClipImageByLayout({repeatX: repeatX = "no-repeat", repeatY: repeatY = "no-repeat", imageMode: imageMode, imageScale: imageScale = 1, imageOffsetX: imageOffsetX = 0, imageOffsetY: imageOffsetY = 0, imagePosition: imagePosition = "top-left"}) {
    const {repeatMode: repeatMode, sizingMode: sizingMode} = resolveImageMode({
        repeatX: repeatX,
        repeatY: repeatY,
        imageMode: imageMode
    });
    return "no-repeat" === repeatMode && ("cover" === sizingMode || "auto" === sizingMode || 1 !== imageScale || 0 !== imageOffsetX || 0 !== imageOffsetY);
}

function drawImageWithLayout(context, data, x, y, width, height, {repeatX: repeatX = "no-repeat", repeatY: repeatY = "no-repeat", imageMode: imageMode, imageScale: imageScale = 1, imageOffsetX: imageOffsetX = 0, imageOffsetY: imageOffsetY = 0, imagePosition: imagePosition = "top-left"}) {
    const {repeatMode: repeatMode, sizingMode: sizingMode} = resolveImageMode({
        repeatX: repeatX,
        repeatY: repeatY,
        imageMode: imageMode
    }), imageBackgroundParams = "no-repeat" === repeatMode ? resolveBackgroundParamsByImageSizing(sizingMode) : {
        backgroundMode: repeatMode,
        backgroundFit: !1,
        backgroundKeepAspectRatio: !1
    };
    (0, base_contribution_render_1.drawBackgroundImage)(context, data, {
        x1: x,
        y1: y,
        x2: x + width,
        y2: y + height,
        width: () => width,
        height: () => height
    }, {
        backgroundMode: imageBackgroundParams.backgroundMode,
        backgroundFit: imageBackgroundParams.backgroundFit,
        backgroundKeepAspectRatio: imageBackgroundParams.backgroundKeepAspectRatio,
        backgroundScale: imageScale,
        backgroundOffsetX: imageOffsetX,
        backgroundOffsetY: imageOffsetY,
        backgroundPosition: imagePosition
    });
}

exports.resolveBackgroundParamsByImageSizing = resolveBackgroundParamsByImageSizing, 
exports.resolveImageRepeatMode = resolveImageRepeatMode, exports.shouldClipImageByLayout = shouldClipImageByLayout, 
exports.drawImageWithLayout = drawImageWithLayout;

let DefaultCanvasImageRender = class extends base_render_1.BaseRender {
    constructor(graphicRenderContributions) {
        super(), this.graphicRenderContributions = graphicRenderContributions, this.numberType = constants_2.IMAGE_NUMBER_TYPE, 
        this.builtinContributions = [ contributions_1.defaultImageRenderContribution, contributions_1.defaultImageBackgroundRenderContribution ], 
        this.init(graphicRenderContributions);
    }
    drawShape(image, context, x, y, drawContext, params, fillCb, strokeCb) {
        const imageAttribute = (0, theme_1.getTheme)(image).image, {repeatX: repeatX = imageAttribute.repeatX, repeatY: repeatY = imageAttribute.repeatY, x: originX = imageAttribute.x, y: originY = imageAttribute.y, cornerRadius: cornerRadius = imageAttribute.cornerRadius, fillStrokeOrder: fillStrokeOrder = imageAttribute.fillStrokeOrder, cornerType: cornerType = imageAttribute.cornerType, imageMode: imageMode = imageAttribute.imageMode, imageScale: imageScale = imageAttribute.imageScale, imageOffsetX: imageOffsetX = imageAttribute.imageOffsetX, imageOffsetY: imageOffsetY = imageAttribute.imageOffsetY, imagePosition: imagePosition = imageAttribute.imagePosition, image: url} = image.attribute, data = this.valid(image, imageAttribute, fillCb);
        if (!data) return;
        const {fVisible: fVisible, sVisible: sVisible, doFill: doFill, doStroke: doStroke} = data;
        if (!url || !image.resources) return;
        const res = image.resources.get(url);
        if ("success" !== res.state) return;
        const width = image.width, height = image.height;
        context.beginPath();
        let needCornerClip = !1;
        0 === cornerRadius || (0, vutils_1.isArray)(cornerRadius) && cornerRadius.every((num => 0 === num)) ? context.rect(x, y, width, height) : ((0, 
        rect_1.createRectPath)(context, x, y, width, height, cornerRadius, "bevel" !== cornerType), 
        needCornerClip = !0), context.setShadowBlendStyle && context.setShadowBlendStyle(image, image.attribute, imageAttribute);
        const _runFill = () => {
            doFill && (fillCb ? fillCb(context, image.attribute, imageAttribute) : fVisible && (context.setCommonStyle(image, image.attribute, x, y, imageAttribute), 
            drawImageWithLayout(context, res.data, x, y, width, height, {
                repeatX: repeatX,
                repeatY: repeatY,
                imageMode: imageMode,
                imageScale: imageScale,
                imageOffsetX: imageOffsetX,
                imageOffsetY: imageOffsetY,
                imagePosition: imagePosition
            })));
        }, _runStroke = () => {
            doStroke && (strokeCb ? strokeCb(context, image.attribute, imageAttribute) : sVisible && (context.setStrokeStyle(image, image.attribute, originX - x, originY - y, imageAttribute), 
            context.stroke()));
        }, needLayoutClip = shouldClipImageByLayout({
            repeatX: repeatX,
            repeatY: repeatY,
            imageMode: imageMode,
            imageScale: imageScale,
            imageOffsetX: imageOffsetX,
            imageOffsetY: imageOffsetY,
            imagePosition: imagePosition
        }), needClip = needCornerClip || needLayoutClip;
        fillStrokeOrder ? (_runStroke(), needClip && (context.save(), context.clip()), this.beforeRenderStep(image, context, x, y, doFill, !1, fVisible, !1, imageAttribute, drawContext, fillCb), 
        _runFill(), needClip && context.restore()) : (needClip && (context.save(), context.clip()), 
        this.beforeRenderStep(image, context, x, y, doFill, !1, fVisible, !1, imageAttribute, drawContext, fillCb), 
        _runFill(), needClip && context.restore(), _runStroke()), this.afterRenderStep(image, context, x, y, doFill, !1, fVisible, !1, imageAttribute, drawContext, fillCb);
    }
    draw(image, renderService, drawContext) {
        const {image: url} = image.attribute;
        if (!url || !image.resources) return;
        const res = image.resources.get(url);
        if ("loading" === res.state && (0, vutils_1.isString)(url)) return void loader_1.ResourceLoader.improveImageLoading(url);
        if ("success" !== res.state) return;
        const {context: context} = renderService.drawParams;
        if (!context) return;
        const imageAttribute = (0, theme_1.getTheme)(image).image;
        this._draw(image, imageAttribute, !1, drawContext);
    }
};

DefaultCanvasImageRender = __decorate([ (0, inversify_lite_1.injectable)(), __param(0, (0, 
inversify_lite_1.inject)(contribution_provider_1.ContributionProvider)), __param(0, (0, 
inversify_lite_1.named)(constants_1.ImageRenderContribution)), __metadata("design:paramtypes", [ Object ]) ], DefaultCanvasImageRender), 
exports.DefaultCanvasImageRender = DefaultCanvasImageRender;
