// @visactor/vchart-extension/esm/charts/timeline/timeline-transformer.js
import { BaseChartSpecTransformer } from "@visactor/vchart";

import { isNil, merge, isObject } from "@visactor/vutils";

export class TimelineChartSpecTransformer extends BaseChartSpecTransformer {
    _getDefaultSeriesSpec(spec) {
        return super._getDefaultSeriesSpec(spec, [ "timeField", "eventField", "seriesField", "titleField", "subTitleField", "iconField", "dot", "title", "subTitle", "icon", "line", "arrow", "labelPosition", "sortDataByAxis" ]);
    }
    _setDefaultXAxisSpec(spec) {
        return {
            type: "band",
            orient: "bottom",
            label: {
                visible: !1
            },
            tick: {
                visible: !1
            },
            grid: {
                visible: !1
            },
            domainLine: {
                visible: !1
            },
            paddingInner: 0,
            paddingOuter: 0
        };
    }
    _setDefaultYAxisSpec(spec) {
        return {
            type: "band",
            inverse: !0,
            orient: "left",
            label: {
                visible: !1
            },
            tick: {
                visible: !1
            },
            grid: {
                visible: !1
            },
            domainLine: {
                visible: !1
            },
            paddingInner: 0,
            paddingOuter: 0
        };
    }
    _transformAxisSpec(spec) {
        spec.axes || (spec.axes = []);
        const haxAxes = {
            x: !1,
            y: !1
        };
        spec.axes.forEach((axis => {
            const {orient: orient} = axis;
            let defaultSpec = null;
            "top" !== orient && "bottom" !== orient || (haxAxes.x = !0, defaultSpec = this._setDefaultXAxisSpec(spec)), 
            "left" !== orient && "right" !== orient || (haxAxes.y = !0, defaultSpec = this._setDefaultYAxisSpec(spec)), 
            defaultSpec && Object.keys(defaultSpec).forEach((key => {
                axis[key] = isObject(axis[key]) ? merge(defaultSpec[key], axis[key]) : isNil(axis[key]) ? defaultSpec[key] : axis[key];
            }));
        })), haxAxes.x || spec.axes.push(this._setDefaultXAxisSpec(spec)), haxAxes.y || spec.axes.push(this._setDefaultYAxisSpec(spec));
    }
    transformSpec(spec) {
        var _a, _b;
        super.transformSpec(spec), this.transformSeriesSpec(spec), this._transformAxisSpec(spec);
        const direction = null !== (_a = spec.direction) && void 0 !== _a ? _a : "horizontal";
        null === (_b = spec.series) || void 0 === _b || _b.forEach((seriesSpec => {
            "vertical" === direction ? (seriesSpec.timeField && !seriesSpec.yField && (seriesSpec.yField = seriesSpec.timeField), 
            seriesSpec.xField || (seriesSpec.xField = seriesSpec.seriesField || "__vchart_timeline_dummy__")) : (seriesSpec.timeField && !seriesSpec.xField && (seriesSpec.xField = seriesSpec.timeField), 
            seriesSpec.yField || (seriesSpec.yField = seriesSpec.seriesField || "__vchart_timeline_dummy__"));
        }));
    }
}
