// @visactor/vchart-extension/cjs/components/regression-line/regression-line.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerRegressionLine = exports.RegressionLine = exports.REGRESSION_LINE = void 0;

const vchart_1 = require("@visactor/vchart"), vchart_2 = require("@visactor/vchart"), vchart_3 = require("@visactor/vchart");

exports.REGRESSION_LINE = "regressionLine";

class RegressionLine extends vchart_1.AbstractComponent {
    constructor() {
        super(...arguments), this.name = exports.REGRESSION_LINE;
    }
    render() {
        this.removeAllChild();
        const {data: data, line: line = {}, label: label, name: name = "regression-line", confidenceInterval: confidenceInterval} = this.attribute;
        (0, vchart_2.isEmpty)(data) || data.forEach((d => {
            const {color: color} = d;
            if (d.area && !1 !== (null == confidenceInterval ? void 0 : confidenceInterval.visible)) {
                const areaShape = (0, vchart_1.createArea)(Object.assign(Object.assign({
                    points: d.area
                }, (0, vchart_2.isValid)(color) ? {
                    fill: color,
                    fillOpacity: .12
                } : null), null == confidenceInterval ? void 0 : confidenceInterval.style));
                areaShape.name = "scatter-regression-area", this.add(areaShape);
            }
            if (d.line && !1 !== (null == line ? void 0 : line.visible)) {
                const lineShape = (0, vchart_1.createLine)(Object.assign(Object.assign({
                    points: d.line,
                    lineWidth: 1
                }, (0, vchart_2.isValid)(color) ? {
                    stroke: color
                } : null), null == line ? void 0 : line.style));
                lineShape.name = `${name}-curve`, this.add(lineShape);
            }
            const lastPoint = (0, vchart_2.last)(d.line);
            if (label && !1 !== label.visible && label.text && lastPoint) {
                const tag = (0, vchart_1.createText)(Object.assign(Object.assign(Object.assign({}, lastPoint), {
                    text: label.text,
                    textAlign: "end",
                    textBaseline: "middle"
                }), label.style));
                tag.name = `${name}-label`, this.add(tag);
            }
        }));
    }
}

exports.RegressionLine = RegressionLine;

const registerRegressionLine = () => {
    vchart_3.Factory.registerGraphicComponent(exports.REGRESSION_LINE, (attrs => new RegressionLine(attrs)));
};

exports.registerRegressionLine = registerRegressionLine;
