// @code-inspector/esbuild/dist/index.mjs
var f = (a, o, n) => new Promise((p, i) => {
  var c = (t) => {
    try {
      r(n.next(t));
    } catch (s) {
      i(s);
    }
  }, e = (t) => {
    try {
      r(n.throw(t));
    } catch (s) {
      i(s);
    }
  }, r = (t) => t.done ? p(t.value) : Promise.resolve(t.value).then(c, e);
  r((n = n.apply(a, o)).next());
});
import { isDev as g, getMappingFilePath as y, isExcludedFile as C, getCodeWithWebComponent as x, isJsTypeFile as T, transformCode as d, parseSFC as v } from "@code-inspector/core";
import F from "fs";
import j from "path";
const P = "@code-inspector/esbuild";
function E(a) {
  return {
    name: P,
    setup(o) {
      if (a.close || !g(a.dev, !1))
        return;
      const n = {
        port: 0,
        entry: "",
        output: a.output
      }, { escapeTags: p = [] } = a, i = /* @__PURE__ */ new Map();
      o.onLoad(
        { filter: a.match || /\.(jsx|tsx|js|ts|mjs|mts)?$/ },
        (c) => f(this, null, function* () {
          let e = c.path;
          e = y(e, a.mappings);
          let r = yield F.promises.readFile(e, "utf8"), t = i.get(e);
          if (!t || t.originCode !== r) {
            let s = r;
            if (C(e, a))
              return s;
            s = yield x({
              options: a,
              file: e,
              code: s,
              record: n
            });
            let l = "";
            if (T(e) ? l = "jsx" : e.endsWith(".svelte") && (l = "svelte"), l)
              s = d({
                content: s,
                filePath: e,
                fileType: l,
                escapeTags: p,
                pathType: a.pathType
              });
            else if (e.endsWith(".vue")) {
              l = "vue";
              const { descriptor: u } = v(s, {
                sourceMap: !1
              }), h = d({
                content: u.template.content,
                filePath: e,
                fileType: l,
                escapeTags: p,
                pathType: a.pathType
              });
              s = s.replace(u.template.content, h);
            }
            const m = j.extname(e).replace(".", "");
            t = { originCode: r, output: { contents: s, loader: m } }, i.set(e, t);
          }
          return t.output;
        })
      );
    }
  };
}
export {
  E as EsbuildCodeInspectorPlugin
};
