// auto-skeleton-react/dist/auto-skeleton-react.js
import { jsx as p, jsxs as Ae } from "react/jsx-runtime";
import { useMemo as Re, useRef as Ce, useState as Te, useLayoutEffect as me } from "react";
function xe(e) {
  var r;
  return e ? (r = e.hasAttribute) != null && r.call(e, "data-no-skeleton") ? !0 : xe(e.parentElement) : !1;
}
function ve(e, r, i = 0) {
  if (i > r.maxDepth) return null;
  if (xe(e)) {
    const c = e.getBoundingClientRect(), l = window.getComputedStyle(e), o = {};
    for (const f of e.attributes)
      o[f.name] = f.value;
    return {
      rect: c,
      style: l,
      tagName: e.tagName,
      textContent: e.textContent,
      attributes: o,
      children: [],
      passthrough: !0,
      passthroughHtml: e.outerHTML
    };
  }
  for (const c of r.ignoreSelectors)
    if (e.matches(c)) return null;
  const t = e.getBoundingClientRect(), s = window.getComputedStyle(e), y = e.tagName === "IMG", n = y ? e : null, S = n == null ? void 0 : n.getAttribute("src"), m = (n == null ? void 0 : n.src) || "", u = (n == null ? void 0 : n.naturalWidth) === 0 || (n == null ? void 0 : n.naturalHeight) === 0, a = !S || S === "", w = m === window.location.href || m.endsWith("/"), R = y && (a || w || u);
  let v = t;
  if (R && n) {
    let c = 0, l = 0;
    const o = e.parentElement, f = o ? window.getComputedStyle(o) : null;
    let b = 0;
    if (o) {
      const d = o.getBoundingClientRect(), H = parseFloat((f == null ? void 0 : f.paddingLeft) || "0"), B = parseFloat((f == null ? void 0 : f.paddingRight) || "0");
      b = d.width - H - B;
    }
    const x = n.style.width, T = n.style.height;
    if (T && !T.includes("%")) {
      const d = parseFloat(T);
      !isNaN(d) && d > 0 && (l = d);
    }
    if (l === 0) {
      const d = parseFloat(s.height);
      !isNaN(d) && d > 50 && (l = d);
    }
    if (x)
      if (x.includes("%")) {
        const d = parseFloat(x) / 100;
        !isNaN(d) && b > 0 && (c = b * d);
      } else {
        const d = parseFloat(x);
        !isNaN(d) && d > 0 && (c = d);
      }
    if (c === 0 && b > 0 && (c = b), c === 0) {
      const d = n.getAttribute("width");
      d && (c = parseFloat(d));
    }
    if (l === 0) {
      const d = n.getAttribute("height");
      d && (l = parseFloat(d));
    }
    c > 0 && l > 0 && (v = {
      x: t.x,
      y: t.y,
      width: c,
      height: l,
      top: t.top,
      right: t.right,
      bottom: t.bottom,
      left: t.left,
      toJSON: t.toJSON.bind(t)
    });
  }
  if (v.width === 0 || v.height === 0 || s.display === "none" || s.visibility === "hidden")
    return null;
  const k = {};
  for (const c of e.attributes)
    k[c.name] = c.value;
  const A = [];
  for (const c of e.children)
    if (c instanceof HTMLElement) {
      const l = ve(c, r, i + 1);
      l && A.push(l);
    }
  return {
    rect: v,
    style: s,
    tagName: e.tagName,
    textContent: e.textContent,
    attributes: k,
    children: A
  };
}
const be = ["TABLE", "THEAD", "TBODY", "TFOOT", "TR"], we = ["TH", "TD"];
function Ne(e) {
  var S, m;
  const { tagName: r, textContent: i, children: t, style: s, rect: y } = e;
  if (y.width * y.height < 100 || be.includes(r) || we.includes(r)) return !1;
  if (r.match(/^(IMG|BUTTON|INPUT|TEXTAREA|SELECT|SVG)$/) || i != null && i.trim() && t.length === 0) return !0;
  if (t.length === 1 && (i != null && i.trim())) {
    const u = t[0];
    if ((S = u.textContent) != null && S.trim() && u.children.length === 0) return !0;
  }
  return t.length > 1 || (m = s.display) != null && m.match(/flex|grid/) ? !1 : t.length === 0;
}
function L(e, r) {
  var c;
  if (e.passthrough)
    return {
      type: "passthrough",
      rect: {
        x: e.rect.x,
        y: e.rect.y,
        width: e.rect.width,
        height: e.rect.height
      },
      passthrough: !0,
      passthroughHtml: e.passthroughHtml
    };
  const { rect: i, style: t, tagName: s, textContent: y, attributes: n, children: S } = e, m = n["data-skeleton-role"];
  if (m) {
    const l = m;
    let o = 1;
    if (l === "text") {
      const f = parseFloat(t.lineHeight), b = parseFloat(t.fontSize), x = f || b * 1.2;
      x > 0 && i.height > x && (o = Math.ceil(i.height / x));
    }
    return {
      type: l,
      rect: {
        x: i.x,
        y: i.y,
        width: i.width,
        height: i.height
      },
      borderRadius: parseFloat(t.borderRadius) || r.borderRadius,
      lines: l === "text" ? o : void 0,
      preservedStyles: {
        display: t.display,
        margin: t.margin,
        padding: t.padding,
        lineHeight: t.lineHeight,
        flex: t.flex,
        flexGrow: t.flexGrow,
        flexShrink: t.flexShrink,
        flexBasis: t.flexBasis,
        alignSelf: t.alignSelf,
        justifySelf: t.justifySelf,
        gridColumn: t.gridColumn,
        gridRow: t.gridRow,
        gridArea: t.gridArea,
        verticalAlign: t.verticalAlign
      }
    };
  }
  if (be.includes(s)) {
    const l = S.map((o) => L(o, r)).filter((o) => o.type !== "skip");
    return {
      type: s.toLowerCase(),
      tagName: s,
      rect: {
        x: i.x,
        y: i.y,
        width: i.width,
        height: i.height
      },
      children: l,
      preservedStyles: {
        display: t.display,
        margin: t.margin,
        padding: t.padding,
        border: t.border,
        borderBottom: t.borderBottom,
        backgroundColor: t.backgroundColor,
        width: t.width,
        borderCollapse: t.borderCollapse,
        tableLayout: t.tableLayout
      }
    };
  }
  if (we.includes(s)) {
    const l = S.map((o) => L(o, r)).filter((o) => o.type !== "skip");
    return {
      type: s.toLowerCase(),
      tagName: s,
      rect: {
        x: i.x,
        y: i.y,
        width: i.width,
        height: i.height
      },
      children: l,
      preservedStyles: {
        display: t.display,
        margin: t.margin,
        padding: t.padding,
        border: t.border,
        borderBottom: t.borderBottom,
        backgroundColor: t.backgroundColor,
        textAlign: t.textAlign,
        verticalAlign: t.verticalAlign,
        width: t.width
      }
    };
  }
  if (!Ne(e)) {
    const l = S.map((o) => L(o, r)).filter((o) => o.type !== "skip");
    return {
      type: "container",
      rect: {
        x: i.x,
        y: i.y,
        width: i.width,
        height: i.height
      },
      display: t.display,
      gap: t.gap,
      borderRadius: parseFloat(t.borderRadius) || 0,
      children: l,
      preservedStyles: {
        display: t.display,
        margin: t.margin,
        padding: t.padding,
        border: t.border,
        boxSizing: t.boxSizing,
        flex: t.flex,
        flexGrow: t.flexGrow,
        flexShrink: t.flexShrink,
        flexBasis: t.flexBasis,
        alignSelf: t.alignSelf,
        justifySelf: t.justifySelf,
        justifyContent: t.justifyContent,
        alignItems: t.alignItems,
        flexDirection: t.flexDirection,
        minHeight: t.minHeight,
        gridTemplateColumns: t.gridTemplateColumns,
        gridTemplateRows: t.gridTemplateRows,
        gridColumn: t.gridColumn,
        gridRow: t.gridRow,
        gridArea: t.gridArea
      }
    };
  }
  const u = i.width * i.height, a = {
    text: 0,
    image: 0,
    icon: 0,
    button: 0,
    input: 0,
    container: 0,
    skip: 0
  };
  y && y.trim().length > 0 && (a.text += 40), i.height < r.minTextHeight * 3 && i.height >= r.minTextHeight && (a.text += 30), s.match(/^(P|SPAN|H[1-6]|LABEL|A|DIV)$/) && (a.text += 20);
  const w = parseFloat(t.fontSize);
  w > 0 && w < 100 && (a.text += 20), s === "IMG" && (a.image += 100), n.role === "img" && (a.image += 60), t.backgroundImage !== "none" && (a.image += 50), u > r.minImageSize ** 2 && (a.image += 30), s === "SVG" && (a.icon += 70);
  const R = i.width / i.height;
  u < r.iconMaxSize ** 2 && R > 0.8 && R < 1.2 && (a.icon += 40), s === "BUTTON" && (a.button += 80), n.role === "button" && (a.button += 60), t.cursor === "pointer" && u < 2e4 && (a.button += 30), s.match(/^(INPUT|TEXTAREA|SELECT)$/) && (a.input += 80), n.contenteditable && (a.input += 50), S.length > 1 && (a.container += 50), (c = t.display) != null && c.match(/flex|grid/) && (a.container += 30), u < 100 && (a.skip += 50), (i.height < 5 || i.width < 5) && (a.skip += 40);
  const v = Object.entries(a).filter(
    ([l]) => l !== "container"
  ).reduce((l, o) => o[1] > l[1] ? o : l), k = v[1] > 30 ? v[0] : "text";
  let A = 1;
  if (k === "text") {
    const l = parseFloat(t.lineHeight), o = parseFloat(t.fontSize), f = l || o * 1.2;
    f > 0 && i.height > f && (A = Math.ceil(i.height / f));
  }
  return {
    type: k,
    rect: {
      x: i.x,
      y: i.y,
      width: i.width,
      height: i.height
    },
    borderRadius: parseFloat(t.borderRadius) || r.borderRadius,
    lines: k === "text" ? A : void 0,
    preservedStyles: {
      display: t.display,
      margin: t.margin,
      padding: t.padding,
      lineHeight: t.lineHeight,
      flex: t.flex,
      flexGrow: t.flexGrow,
      flexShrink: t.flexShrink,
      flexBasis: t.flexBasis,
      alignSelf: t.alignSelf,
      justifySelf: t.justifySelf,
      gridColumn: t.gridColumn,
      gridRow: t.gridRow,
      gridArea: t.gridArea,
      verticalAlign: t.verticalAlign
    }
  };
}
function He({
  blueprint: e,
  baseColor: r,
  highlightColor: i,
  animation: t
}) {
  return C(e, r, i, t);
}
function C(e, r, i, t, s) {
  var a, w, R, v, k, A, c, l, o, f, b, x, T, d, H, B, F, I, z, j, E, D, M, O, W, _, P, U, V, $, J, X, q, Y, K, Q, Z, ee, te, ie, re, se, le, ne, ae, oe, de, ce, he, pe, fe, ge, ue, ye, Se;
  if (e.type === "passthrough")
    return e.passthroughHtml ? /* @__PURE__ */ p(
      "div",
      {
        style: { display: "contents" },
        dangerouslySetInnerHTML: { __html: e.passthroughHtml }
      },
      s
    ) : null;
  if (e.type === "skip") return null;
  if (e.type === "table")
    return /* @__PURE__ */ p(
      "table",
      {
        style: {
          width: ((a = e.preservedStyles) == null ? void 0 : a.width) || "100%",
          borderCollapse: (w = e.preservedStyles) == null ? void 0 : w.borderCollapse,
          tableLayout: (R = e.preservedStyles) == null ? void 0 : R.tableLayout
        },
        children: (v = e.children) == null ? void 0 : v.map(
          (h, g) => C(h, r, i, t, g)
        )
      },
      s
    );
  if (e.type === "thead")
    return /* @__PURE__ */ p("thead", { children: (k = e.children) == null ? void 0 : k.map(
      (h, g) => C(h, r, i, t, g)
    ) }, s);
  if (e.type === "tbody")
    return /* @__PURE__ */ p("tbody", { children: (A = e.children) == null ? void 0 : A.map(
      (h, g) => C(h, r, i, t, g)
    ) }, s);
  if (e.type === "tr")
    return /* @__PURE__ */ p("tr", { children: (c = e.children) == null ? void 0 : c.map(
      (h, g) => C(h, r, i, t, g)
    ) }, s);
  if (e.type === "th" || e.type === "td") {
    const h = e.type, g = {
      padding: (l = e.preservedStyles) == null ? void 0 : l.padding,
      borderBottom: (o = e.preservedStyles) == null ? void 0 : o.borderBottom,
      backgroundColor: (f = e.preservedStyles) == null ? void 0 : f.backgroundColor,
      textAlign: (b = e.preservedStyles) == null ? void 0 : b.textAlign,
      verticalAlign: (x = e.preservedStyles) == null ? void 0 : x.verticalAlign,
      width: (T = e.preservedStyles) == null ? void 0 : T.width
    };
    return e.children && e.children.length > 0 ? /* @__PURE__ */ p(h, { style: g, children: e.children.map(
      (N, ke) => C(N, r, i, t, ke)
    ) }, s) : /* @__PURE__ */ p(h, { style: g, children: /* @__PURE__ */ p("div", { style: {
      width: "80%",
      height: 16,
      borderRadius: 4,
      backgroundColor: r,
      animation: t === "pulse" ? "skeleton-pulse 2s ease-in-out infinite" : void 0
    } }) }, s);
  }
  if (e.type === "container" && e.children && e.children.length > 0) {
    const h = ((d = e.preservedStyles) == null ? void 0 : d.border) || "", g = h && !h.includes("0px") && !h.startsWith("none");
    return /* @__PURE__ */ p(
      "div",
      {
        style: {
          display: e.display || "block",
          gap: e.gap,
          width: e.rect.width,
          // Don't set fixed height - let container size naturally from children
          // This avoids issues with incorrect height measurement from alt text
          boxSizing: "border-box",
          margin: (H = e.preservedStyles) == null ? void 0 : H.margin,
          padding: (B = e.preservedStyles) == null ? void 0 : B.padding,
          border: g ? (F = e.preservedStyles) == null ? void 0 : F.border : void 0,
          borderRadius: e.borderRadius,
          flex: (I = e.preservedStyles) == null ? void 0 : I.flex,
          flexGrow: (z = e.preservedStyles) != null && z.flexGrow ? Number(e.preservedStyles.flexGrow) : void 0,
          flexShrink: (j = e.preservedStyles) != null && j.flexShrink ? Number(e.preservedStyles.flexShrink) : void 0,
          flexBasis: (E = e.preservedStyles) == null ? void 0 : E.flexBasis,
          alignSelf: (D = e.preservedStyles) == null ? void 0 : D.alignSelf,
          justifySelf: (M = e.preservedStyles) == null ? void 0 : M.justifySelf,
          justifyContent: (O = e.preservedStyles) == null ? void 0 : O.justifyContent,
          alignItems: (W = e.preservedStyles) == null ? void 0 : W.alignItems,
          flexDirection: (_ = e.preservedStyles) == null ? void 0 : _.flexDirection,
          minHeight: (P = e.preservedStyles) == null ? void 0 : P.minHeight,
          // Preserve measured grid template so responsive breakpoints (e.g. md:grid-cols-*)
          // are kept exactly during skeleton rendering.
          gridTemplateColumns: (U = e.preservedStyles) == null ? void 0 : U.gridTemplateColumns,
          gridTemplateRows: (V = e.preservedStyles) == null ? void 0 : V.gridTemplateRows,
          gridColumn: ($ = e.preservedStyles) == null ? void 0 : $.gridColumn,
          gridRow: (J = e.preservedStyles) == null ? void 0 : J.gridRow,
          gridArea: (X = e.preservedStyles) == null ? void 0 : X.gridArea
        },
        children: e.children.map(
          (G, N) => C(G, r, i, t, N)
        )
      },
      s
    );
  }
  if (e.type === "image") {
    const h = {
      width: e.rect.width,
      height: e.rect.height,
      borderRadius: e.borderRadius,
      backgroundColor: r,
      display: "block",
      boxSizing: "border-box",
      margin: (q = e.preservedStyles) == null ? void 0 : q.margin
    };
    return t === "pulse" && (h.animation = "skeleton-pulse 2s ease-in-out infinite"), /* @__PURE__ */ p("div", { style: h, className: "skeleton-image" }, s);
  }
  if (e.type === "button" || e.type === "input") {
    const h = {
      // Lock exact dimensions
      width: e.rect.width,
      height: e.rect.height,
      minWidth: e.rect.width,
      minHeight: e.rect.height,
      boxSizing: "border-box",
      // Preserve flex-item properties
      margin: (Y = e.preservedStyles) == null ? void 0 : Y.margin,
      flex: (K = e.preservedStyles) == null ? void 0 : K.flex,
      flexGrow: (Q = e.preservedStyles) != null && Q.flexGrow ? Number(e.preservedStyles.flexGrow) : void 0,
      flexShrink: 0,
      // Prevent flex compression
      flexBasis: (Z = e.preservedStyles) == null ? void 0 : Z.flexBasis,
      alignSelf: (ee = e.preservedStyles) == null ? void 0 : ee.alignSelf,
      justifySelf: (te = e.preservedStyles) == null ? void 0 : te.justifySelf,
      // Grid properties
      gridColumn: (ie = e.preservedStyles) == null ? void 0 : ie.gridColumn,
      gridRow: (re = e.preservedStyles) == null ? void 0 : re.gridRow,
      gridArea: (se = e.preservedStyles) == null ? void 0 : se.gridArea,
      // Visual
      borderRadius: e.borderRadius,
      backgroundColor: r,
      // Preserve inline nature
      display: "inline-block",
      verticalAlign: ((le = e.preservedStyles) == null ? void 0 : le.verticalAlign) || "middle"
    };
    return t === "pulse" && (h.animation = "skeleton-pulse 2s ease-in-out infinite"), /* @__PURE__ */ p("div", { style: h, className: `skeleton-${e.type}` }, s);
  }
  const n = {
    // Only margin for spacing between siblings
    margin: (ne = e.preservedStyles) == null ? void 0 : ne.margin,
    // Flex-item properties
    flex: (ae = e.preservedStyles) == null ? void 0 : ae.flex,
    flexGrow: (oe = e.preservedStyles) != null && oe.flexGrow ? Number(e.preservedStyles.flexGrow) : void 0,
    flexShrink: (de = e.preservedStyles) != null && de.flexShrink ? Number(e.preservedStyles.flexShrink) : void 0,
    flexBasis: (ce = e.preservedStyles) == null ? void 0 : ce.flexBasis,
    alignSelf: (he = e.preservedStyles) == null ? void 0 : he.alignSelf,
    // Grid-item properties
    gridColumn: (pe = e.preservedStyles) == null ? void 0 : pe.gridColumn,
    gridRow: (fe = e.preservedStyles) == null ? void 0 : fe.gridRow,
    gridArea: (ge = e.preservedStyles) == null ? void 0 : ge.gridArea
  }, S = ((ue = e.preservedStyles) == null ? void 0 : ue.display) === "inline" ? "inline-block" : "block", m = {
    width: e.rect.width,
    height: e.rect.height,
    borderRadius: e.borderRadius,
    backgroundColor: r,
    display: S,
    verticalAlign: ((ye = e.preservedStyles) == null ? void 0 : ye.verticalAlign) || "middle"
  };
  if (t === "pulse" && (m.animation = "skeleton-pulse 2s ease-in-out infinite"), e.type === "text" && e.lines && e.lines > 1) {
    const h = parseFloat(((Se = e.preservedStyles) == null ? void 0 : Se.lineHeight) || "1.2"), g = e.rect.height / e.lines;
    return /* @__PURE__ */ p(
      "div",
      {
        style: {
          ...n,
          display: "flex",
          flexDirection: "column",
          gap: h > g ? h - g : 4
        },
        children: Array.from({ length: e.lines }).map((G, N) => /* @__PURE__ */ p(
          "div",
          {
            style: {
              ...m,
              height: g,
              width: N === e.lines - 1 ? e.rect.width * 0.7 : e.rect.width
              // Last line shorter
            },
            className: `skeleton-${e.type}`
          },
          N
        ))
      },
      s
    );
  }
  const u = `skeleton-${e.type}`;
  return /* @__PURE__ */ p("div", { style: n, children: /* @__PURE__ */ p("div", { style: m, className: u }) }, s);
}
const Be = {
  animation: "pulse",
  baseColor: "#e0e0e0",
  highlightColor: "#f5f5f5",
  borderRadius: 4,
  minTextHeight: 12,
  minImageSize: 32,
  iconMaxSize: 48,
  maxDepth: 10,
  ignoreSelectors: [".no-skeleton", "[data-skeleton-ignore]"]
};
function Fe({
  loading: e,
  children: r,
  config: i
}) {
  const t = Re(
    () => ({ ...Be, ...i }),
    [i]
  ), s = Ce(null), [y, n] = Te(null);
  return me(() => {
    if (!e || !s.current)
      return;
    const u = requestAnimationFrame(() => {
      if (s.current)
        try {
          const a = ve(s.current, t);
          if (a) {
            const w = L(a, t);
            n(w);
          }
        } catch (a) {
          console.error("AutoSkeleton measurement failed:", a), n(null);
        }
    });
    return () => cancelAnimationFrame(u);
  }, [e, t]), me(() => {
    if (e) return;
    const u = setTimeout(() => {
      n(null);
    }, 300);
    return () => clearTimeout(u);
  }, [e]), /* @__PURE__ */ Ae("div", { style: { position: "relative" }, children: [
    /* @__PURE__ */ p("style", { children: `
        @keyframes skeleton-pulse {
          0%, 100% { opacity: 1; }
          50% { opacity: 0.5; }
        }
        .auto-skeleton-fade {
          transition: opacity 0.3s ease-out;
        }
      ` }),
    /* @__PURE__ */ p(
      "div",
      {
        className: "auto-skeleton-fade",
        style: { opacity: e && y !== null ? 0 : 1 },
        children: r
      }
    ),
    (e || y !== null) && y && /* @__PURE__ */ p(
      "div",
      {
        className: "auto-skeleton-fade",
        style: {
          position: "absolute",
          top: 0,
          left: 0,
          right: 0,
          opacity: e ? 1 : 0,
          pointerEvents: e ? "auto" : "none"
        },
        children: /* @__PURE__ */ p(
          He,
          {
            blueprint: y,
            baseColor: t.baseColor,
            highlightColor: t.highlightColor,
            animation: t.animation
          }
        )
      }
    ),
    e && /* @__PURE__ */ p(
      "div",
      {
        ref: s,
        style: {
          opacity: 0,
          position: "absolute",
          top: 0,
          left: 0,
          right: 0,
          pointerEvents: "none",
          zIndex: -1
        },
        children: r
      }
    )
  ] });
}
export {
  Fe as AutoSkeleton
};
