// @visactor/vchart/cjs/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function(o, m, k, k2) {
    void 0 === k2 && (k2 = k);
    var desc = Object.getOwnPropertyDescriptor(m, k);
    desc && !("get" in desc ? !m.__esModule : desc.writable || desc.configurable) || (desc = {
        enumerable: !0,
        get: function() {
            return m[k];
        }
    }), Object.defineProperty(o, k2, desc);
} : function(o, m, k, k2) {
    void 0 === k2 && (k2 = k), o[k2] = m[k];
}), __exportStar = this && this.__exportStar || function(m, exports) {
    for (var p in m) "default" === p || Object.prototype.hasOwnProperty.call(exports, p) || __createBinding(exports, m, p);
};

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.svgParser = exports.DataSet = exports.DataView = exports.registerShadowRoot = exports.registerRect3d = exports.registerPyramid3d = exports.registerArc3d = exports.registerRect = exports.registerLine = exports.MarkPoint = exports.Segment = exports.AbstractComponent = exports.RotateBySphereAnimate = exports.AnimateExecutor = exports.ACustomAnimate = exports.graphicCreator = exports.registerViewTransform3dPlugin = exports.registerOrthoCamera = exports.registerDirectionalLight = exports.createRect3d = exports.createPyramid3d = exports.createArc3d = exports.createPath = exports.createImage = void 0;

const vchart_all_1 = require("./vchart-all");

exports.default = vchart_all_1.VChart, __exportStar(require("./core"), exports), 
__exportStar(require("./chart"), exports), __exportStar(require("./chart/base"), exports), 
__exportStar(require("./chart/cartesian"), exports), __exportStar(require("./chart/common"), exports), 
__exportStar(require("./chart/stack"), exports), __exportStar(require("./series"), exports), 
__exportStar(require("./mark"), exports), __exportStar(require("./component"), exports), 
__exportStar(require("./layout"), exports), __exportStar(require("./env"), exports), 
__exportStar(require("./event"), exports), __exportStar(require("./plugin/chart"), exports), 
__exportStar(require("./plugin/components/tooltip-handler"), exports), __exportStar(require("./plugin/components/axis-sync"), exports), 
__exportStar(require("./plugin/other"), exports), __exportStar(require("./plugin/base/base-plugin"), exports), 
__exportStar(require("./vrender-tools"), exports), __exportStar(require("./util/data"), exports), 
__exportStar(require("./util/spec/transform"), exports), __exportStar(require("./util/mark"), exports), 
__exportStar(require("./util/region"), exports), __exportStar(require("./util/scale"), exports), 
__exportStar(require("./constant/data"), exports), __exportStar(require("./component/base"), exports), 
__exportStar(require("./compile/data/compilable-data"), exports), __exportStar(require("./animation"), exports), 
__exportStar(require("./interaction"), exports);

var vrender_core_1 = require("@visactor/vrender-core");

Object.defineProperty(exports, "createImage", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createImage;
    }
}), Object.defineProperty(exports, "createPath", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createPath;
    }
}), Object.defineProperty(exports, "createArc3d", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createArc3d;
    }
}), Object.defineProperty(exports, "createPyramid3d", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createPyramid3d;
    }
}), Object.defineProperty(exports, "createRect3d", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.createRect3d;
    }
}), Object.defineProperty(exports, "registerDirectionalLight", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.registerDirectionalLight;
    }
}), Object.defineProperty(exports, "registerOrthoCamera", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.registerOrthoCamera;
    }
}), Object.defineProperty(exports, "registerViewTransform3dPlugin", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.registerViewTransform3dPlugin;
    }
}), Object.defineProperty(exports, "graphicCreator", {
    enumerable: !0,
    get: function() {
        return vrender_core_1.graphicCreator;
    }
});

var vrender_animate_1 = require("@visactor/vrender-animate");

Object.defineProperty(exports, "ACustomAnimate", {
    enumerable: !0,
    get: function() {
        return vrender_animate_1.ACustomAnimate;
    }
}), Object.defineProperty(exports, "AnimateExecutor", {
    enumerable: !0,
    get: function() {
        return vrender_animate_1.AnimateExecutor;
    }
}), Object.defineProperty(exports, "RotateBySphereAnimate", {
    enumerable: !0,
    get: function() {
        return vrender_animate_1.RotateBySphereAnimate;
    }
});

var vrender_components_1 = require("@visactor/vrender-components");

Object.defineProperty(exports, "AbstractComponent", {
    enumerable: !0,
    get: function() {
        return vrender_components_1.AbstractComponent;
    }
}), Object.defineProperty(exports, "Segment", {
    enumerable: !0,
    get: function() {
        return vrender_components_1.Segment;
    }
}), Object.defineProperty(exports, "MarkPoint", {
    enumerable: !0,
    get: function() {
        return vrender_components_1.MarkPoint;
    }
});

var vrender_kits_1 = require("@visactor/vrender-kits");

Object.defineProperty(exports, "registerLine", {
    enumerable: !0,
    get: function() {
        return vrender_kits_1.registerLine;
    }
}), Object.defineProperty(exports, "registerRect", {
    enumerable: !0,
    get: function() {
        return vrender_kits_1.registerRect;
    }
}), Object.defineProperty(exports, "registerArc3d", {
    enumerable: !0,
    get: function() {
        return vrender_kits_1.registerArc3d;
    }
}), Object.defineProperty(exports, "registerPyramid3d", {
    enumerable: !0,
    get: function() {
        return vrender_kits_1.registerPyramid3d;
    }
}), Object.defineProperty(exports, "registerRect3d", {
    enumerable: !0,
    get: function() {
        return vrender_kits_1.registerRect3d;
    }
}), Object.defineProperty(exports, "registerShadowRoot", {
    enumerable: !0,
    get: function() {
        return vrender_kits_1.registerShadowRoot;
    }
}), __exportStar(require("@visactor/vlayouts"), exports);

var vdataset_1 = require("@visactor/vdataset");

Object.defineProperty(exports, "DataView", {
    enumerable: !0,
    get: function() {
        return vdataset_1.DataView;
    }
}), Object.defineProperty(exports, "DataSet", {
    enumerable: !0,
    get: function() {
        return vdataset_1.DataSet;
    }
}), Object.defineProperty(exports, "svgParser", {
    enumerable: !0,
    get: function() {
        return vdataset_1.svgParser;
    }
}), __exportStar(require("@visactor/vutils"), exports);
