// @visactor/vchart-extension/cjs/charts/timeline/series/event-series.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.registerEventSeries = exports.EventSeries = void 0;

const vchart_1 = require("@visactor/vchart"), constant_1 = require("./constant"), theme_1 = require("./theme"), vutils_1 = require("@visactor/vutils"), eventSeriesMark = Object.assign(Object.assign({}, vchart_1.baseSeriesMark), {
    line: {
        name: "line",
        type: "line"
    },
    dot: {
        name: "dot",
        type: "symbol"
    },
    icon: {
        name: "icon",
        type: "symbol"
    },
    title: {
        name: "title",
        type: "text"
    },
    subTitle: {
        name: "subTitle",
        type: "text"
    },
    arrow: {
        name: "arrow",
        type: "path"
    }
});

class EventSeries extends vchart_1.CartesianSeries {
    constructor() {
        super(...arguments), this.type = constant_1.EVENT_SERIES_TYPE, this.transformerConstructor = vchart_1.BaseSeriesSpecTransformer, 
        this.coordinate = "cartesian";
    }
    setAttrFromSpec() {
        super.setAttrFromSpec(), this.setSeriesField(this._spec.seriesField), this._timeField = this._spec.timeField, 
        this._eventField = this._spec.eventField, this._subTitleField = this._spec.subTitleField, 
        this._iconField = this._spec.iconField, this._labelPosition = this._spec.labelPosition;
    }
    getDimensionField() {
        return this._timeField ? [ this._timeField ] : [];
    }
    getMeasureField() {
        const fields = [];
        return this._eventField && fields.push(this._eventField), this._subTitleField && fields.push(this._subTitleField), 
        fields;
    }
    initMark() {
        this._axisLineMark = this._createMark(EventSeries.mark.line, {
            isSeriesMark: !0,
            groupKey: this._seriesField
        }), this._arrowMark = this._createMark(EventSeries.mark.arrow, {
            isSeriesMark: !0
        }), this._dotMark = this._createMark(EventSeries.mark.dot, {
            isSeriesMark: !0
        }), this._iconMark = this._createMark(EventSeries.mark.icon, {
            isSeriesMark: !0
        }), this._titleMark = this._createMark(EventSeries.mark.title), this._subTitleMark = this._createMark(EventSeries.mark.subTitle);
    }
    initTooltip() {
        super.initTooltip(), this._dotMark && this._tooltipHelper.activeTriggerSet.mark.add(this._dotMark), 
        this._titleMark && this._tooltipHelper.activeTriggerSet.mark.add(this._titleMark), 
        this._subTitleMark && this._tooltipHelper.activeTriggerSet.mark.add(this._subTitleMark), 
        this._iconMark && this._tooltipHelper.activeTriggerSet.mark.add(this._iconMark), 
        this._arrowMark && this._tooltipHelper.activeTriggerSet.mark.add(this._arrowMark);
    }
    initMarkStyle() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        this._axisLineMark && this.setMarkStyle(this._axisLineMark, {
            points: this._getAxisPoints.bind(this),
            visible: this._isFirstDataInGroup.bind(this),
            stroke: "#c0c3c7",
            lineWidth: 1
        }, vchart_1.STATE_VALUE_ENUM.STATE_NORMAL, vchart_1.AttributeLevel.Series);
        const dotSpec = this._spec.dot, dotSize = "number" == typeof (null === (_a = null == dotSpec ? void 0 : dotSpec.style) || void 0 === _a ? void 0 : _a.size) ? dotSpec.style.size : 8, titleSpec = this._spec.title, titleStyle = null !== (_b = null == titleSpec ? void 0 : titleSpec.style) && void 0 !== _b ? _b : {}, subTitleSpec = this._spec.subTitle, subTitleStyle = null !== (_c = null == subTitleSpec ? void 0 : subTitleSpec.style) && void 0 !== _c ? _c : {}, titleFontSize = "number" == typeof titleStyle.fontSize ? titleStyle.fontSize : 14, labelOffset = dotSize / 2 + (null !== (_d = null == titleSpec ? void 0 : titleSpec.offset) && void 0 !== _d ? _d : 6);
        this._dotMark && this.setMarkStyle(this._dotMark, {
            x: datum => this._getPoint(datum).x,
            y: datum => this._getPoint(datum).y,
            size: dotSize,
            fill: null !== (_f = null === (_e = null == dotSpec ? void 0 : dotSpec.style) || void 0 === _e ? void 0 : _e.fill) && void 0 !== _f ? _f : this.getColorAttribute()
        }, vchart_1.STATE_VALUE_ENUM.STATE_NORMAL, vchart_1.AttributeLevel.Series), this._arrowMark && this.setMarkStyle(this._arrowMark, {
            path: datum => this._getArrowPath(datum),
            fill: null !== (_h = null === (_g = null == dotSpec ? void 0 : dotSpec.style) || void 0 === _g ? void 0 : _g.fill) && void 0 !== _h ? _h : this.getColorAttribute()
        }, vchart_1.STATE_VALUE_ENUM.STATE_NORMAL, vchart_1.AttributeLevel.Series), this._titleMark && this.setMarkStyle(this._titleMark, Object.assign(Object.assign({
            fontSize: 14
        }, titleStyle), {
            x: datum => this._getTitlePosition(datum, labelOffset).x,
            y: datum => this._getTitlePosition(datum, labelOffset).y,
            textAlign: datum => this._getLabelTextAlign(datum),
            textBaseline: datum => this._getLabelTextBaseline(datum, !0),
            text: datum => this._getDatumString(datum, this._eventField)
        }), vchart_1.STATE_VALUE_ENUM.STATE_NORMAL, vchart_1.AttributeLevel.Series), this._subTitleMark && this.setMarkStyle(this._subTitleMark, Object.assign(Object.assign({}, subTitleStyle), {
            x: datum => this._getSubTitlePosition(datum, labelOffset, titleFontSize).x,
            y: datum => this._getSubTitlePosition(datum, labelOffset, titleFontSize).y,
            textAlign: datum => this._getLabelTextAlign(datum),
            textBaseline: datum => this._getLabelTextBaseline(datum, !1),
            text: datum => this._getDatumString(datum, this._subTitleField)
        }), vchart_1.STATE_VALUE_ENUM.STATE_NORMAL, vchart_1.AttributeLevel.Series);
        const iconSpec = this._spec.icon, iconSize = "number" == typeof (null === (_j = null == iconSpec ? void 0 : iconSpec.style) || void 0 === _j ? void 0 : _j.size) ? iconSpec.style.size : 20, iconOffset = ((0, 
        vutils_1.isValid)(null == iconSpec ? void 0 : iconSpec.offset) ? iconSpec.offset + dotSize / 2 : labelOffset) + iconSize / 2;
        this._iconMark && this.setMarkStyle(this._iconMark, {
            x: datum => this._getIconPosition(datum, iconOffset).x,
            y: datum => this._getIconPosition(datum, iconOffset).y,
            size: iconSize,
            fill: null !== (_l = null === (_k = null == iconSpec ? void 0 : iconSpec.style) || void 0 === _k ? void 0 : _k.fill) && void 0 !== _l ? _l : this.getColorAttribute(),
            shape: datum => this._getDatumString(datum, this._iconField) || "circle"
        }, vchart_1.STATE_VALUE_ENUM.STATE_NORMAL, vchart_1.AttributeLevel.Series);
    }
    _getLabelSide(datum) {
        const index = this._getViewDataList().indexOf(datum), position = this._labelPosition;
        if ("vertical" === this.direction) switch (position) {
          case "left":
          default:
            return "primary";

          case "right":
            return "secondary";

          case "left-right":
            return index % 2 == 0 ? "primary" : "secondary";

          case "right-left":
            return index % 2 == 0 ? "secondary" : "primary";
        } else switch (position) {
          case "top":
          default:
            return "primary";

          case "bottom":
            return "secondary";

          case "top-bottom":
            return index % 2 == 0 ? "primary" : "secondary";

          case "bottom-top":
            return index % 2 == 0 ? "secondary" : "primary";
        }
    }
    _getLabelTextAlign(datum) {
        if ("vertical" === this.direction) {
            return "primary" === this._getLabelSide(datum) ? "right" : "left";
        }
        return "center";
    }
    _getLabelTextBaseline(datum, isTitle) {
        if ("vertical" === this.direction) return "middle";
        return "primary" === this._getLabelSide(datum) ? "bottom" : "top";
    }
    _getTitlePosition(datum, offset) {
        const point = this._getPoint(datum), side = this._getLabelSide(datum);
        return "vertical" === this.direction ? {
            x: "primary" === side ? point.x - offset : point.x + offset,
            y: point.y
        } : {
            x: point.x,
            y: "primary" === side ? point.y - offset : point.y + offset
        };
    }
    _getSubTitlePosition(datum, offset, titleFontSize) {
        var _a, _b, _c, _d;
        const point = this._getPoint(datum), side = this._getLabelSide(datum), gap = null !== (_b = null === (_a = this._spec.title) || void 0 === _a ? void 0 : _a.subTitleGap) && void 0 !== _b ? _b : 4;
        if ("vertical" === this.direction) {
            const titleStyle = null !== (_d = null === (_c = this._spec.title) || void 0 === _c ? void 0 : _c.style) && void 0 !== _d ? _d : {}, titleLineHeight = "number" == typeof titleStyle.lineHeight ? titleStyle.lineHeight : 1.2 * titleFontSize;
            return {
                x: "primary" === side ? point.x - offset : point.x + offset,
                y: point.y + titleLineHeight + gap
            };
        }
        return {
            x: point.x,
            y: "primary" === side ? point.y - (offset + titleFontSize + gap) : point.y + (offset + titleFontSize + gap)
        };
    }
    _getIconPosition(datum, offset) {
        const point = this._getPoint(datum), side = this._getLabelSide(datum);
        return "vertical" === this.direction ? {
            x: "primary" === side ? point.x + offset : point.x - offset,
            y: point.y
        } : {
            x: point.x,
            y: "primary" === side ? point.y + offset : point.y - offset
        };
    }
    _getViewDataList() {
        var _a, _b;
        return null !== (_b = null === (_a = this.getViewData()) || void 0 === _a ? void 0 : _a.latestData) && void 0 !== _b ? _b : [];
    }
    _getPoint(datum) {
        if ("vertical" === this.direction) {
            return {
                x: this._getPositionFromAxis(datum, this.getXAxisHelper(), this._seriesField),
                y: this._getPositionFromAxis(datum, this.getYAxisHelper(), this._timeField)
            };
        }
        return {
            x: this._getPositionFromAxis(datum, this.getXAxisHelper(), this._timeField),
            y: this._getPositionFromAxis(datum, this.getYAxisHelper(), this._seriesField)
        };
    }
    _getPositionFromAxis(datum, axisHelper, field) {
        var _a, _b, _c, _d, _e, _f;
        if (!axisHelper) return this._getDefaultPosition(field);
        if (!field || !(field in datum)) return this._getDefaultPosition(field);
        const value = datum[field];
        return "band" !== (null === (_b = null === (_a = axisHelper.getSpec) || void 0 === _a ? void 0 : _a.call(axisHelper)) || void 0 === _b ? void 0 : _b.type) && axisHelper.isContinuous ? null !== (_f = null === (_e = axisHelper.valueToPosition) || void 0 === _e ? void 0 : _e.call(axisHelper, value)) && void 0 !== _f ? _f : this._getDefaultPosition(field) : null !== (_d = null === (_c = axisHelper.dataToPosition) || void 0 === _c ? void 0 : _c.call(axisHelper, [ value ], {
            bandPosition: .5
        })) && void 0 !== _d ? _d : this._getDefaultPosition(field);
    }
    onXAxisHelperUpdate() {
        var _a;
        null === (_a = super.onXAxisHelperUpdate) || void 0 === _a || _a.call(this), this.onMarkPositionUpdate();
    }
    onYAxisHelperUpdate() {
        var _a;
        null === (_a = super.onYAxisHelperUpdate) || void 0 === _a || _a.call(this), this.onMarkPositionUpdate();
    }
    _getDefaultPosition(field) {
        const rect = this._region.getLayoutRect();
        if (!this._seriesField || field === this._seriesField) {
            return "vertical" !== this.direction ? .5 * rect.height : .5 * rect.width;
        }
        return 0;
    }
    _getAxisPoints(datum) {
        const rect = this._region.getLayoutRect();
        if ("vertical" === this.direction) {
            const x = this._getPositionFromAxis(datum, this.getXAxisHelper(), this._seriesField);
            return [ {
                x: x,
                y: 0
            }, {
                x: x,
                y: rect.height
            } ];
        }
        const y = this._getPositionFromAxis(datum, this.getYAxisHelper(), this._seriesField);
        return [ {
            x: 0,
            y: y
        }, {
            x: rect.width,
            y: y
        } ];
    }
    _isFirstDataInGroup(datum) {
        if (!this._seriesField) {
            return 0 === this._getViewDataList().indexOf(datum);
        }
        const data = this._getViewDataList(), categoryValue = datum[this._seriesField];
        return datum === data.find((d => this._seriesField && d[this._seriesField] === categoryValue));
    }
    _getNextDatum(datum) {
        const data = this._getViewDataList(), currentIndex = data.indexOf(datum);
        if (-1 === currentIndex || currentIndex === data.length - 1) return null;
        if (!this._seriesField) return data[currentIndex + 1];
        const categoryValue = datum[this._seriesField];
        for (let i = currentIndex + 1; i < data.length; i++) {
            const nextDatum = data[i];
            if (nextDatum[this._seriesField] === categoryValue) return nextDatum;
        }
        return null;
    }
    _getPreviousDatum(datum) {
        const data = this._getViewDataList(), currentIndex = data.indexOf(datum);
        if (-1 === currentIndex || 0 === currentIndex) return null;
        if (!this._seriesField) return data[currentIndex - 1];
        const categoryValue = datum[this._seriesField];
        for (let i = currentIndex - 1; i >= 0; i--) {
            const prevDatum = data[i];
            if (prevDatum[this._seriesField] === categoryValue) return prevDatum;
        }
        return null;
    }
    _getArrowPath(datum) {
        var _a, _b;
        const point = this._getPoint(datum), nextDatum = this._getNextDatum(datum), prevDatum = this._getPreviousDatum(datum), arrowThickness = null !== (_b = null === (_a = this._spec.arrow) || void 0 === _a ? void 0 : _a.thickness) && void 0 !== _b ? _b : 16, rect = this._region.getLayoutRect();
        if ("vertical" === this.direction) {
            const inverse = this.getYAxisHelper().isInverse(), startPoint = prevDatum ? {
                x: point.x,
                y: (this._getPoint(prevDatum).y + point.y) / 2
            } : {
                x: point.x,
                y: inverse ? 0 : rect.height
            }, endPoint = nextDatum ? {
                x: point.x,
                y: (this._getPoint(nextDatum).y + point.y) / 2
            } : {
                x: point.x,
                y: inverse ? rect.height : 0
            }, tag = inverse ? 1 : -1, arrowHeight = arrowThickness / 3, arrowWidth = arrowThickness / 2;
            return `M ${startPoint.x - arrowWidth} ${startPoint.y} L ${startPoint.x} ${startPoint.y + tag * arrowHeight} L ${startPoint.x + arrowWidth} ${startPoint.y} \n      L ${endPoint.x + arrowWidth} ${endPoint.y - tag * arrowHeight} \n      L ${endPoint.x} ${endPoint.y}\n      L ${endPoint.x - arrowWidth} ${endPoint.y - tag * arrowHeight} Z`;
        }
        const inverse = this.getXAxisHelper().isInverse(), startPoint = prevDatum ? {
            x: (this._getPoint(prevDatum).x + point.x) / 2,
            y: point.y
        } : {
            x: inverse ? rect.width : 0,
            y: point.y
        }, endPoint = nextDatum ? {
            x: (this._getPoint(nextDatum).x + point.x) / 2,
            y: point.y
        } : {
            x: inverse ? 0 : rect.width,
            y: point.y
        }, tag = inverse ? -1 : 1, arrowHeight = arrowThickness / 2, arrowWidth = arrowThickness / 3;
        return `M ${startPoint.x} ${startPoint.y - arrowHeight} L ${startPoint.x + tag * arrowWidth} ${startPoint.y} L ${startPoint.x} ${startPoint.y + arrowHeight} \n      L ${endPoint.x - tag * arrowWidth} ${endPoint.y + arrowHeight} \n      L ${endPoint.x} ${endPoint.y} \n      L ${endPoint.x - tag * arrowWidth} ${endPoint.y - arrowHeight} Z`;
    }
    _getDatumString(datum, field) {
        if (!datum || !field) return "";
        const value = datum[field];
        return "string" == typeof value ? value : null == value ? "" : String(value);
    }
    valueToPosition(timeValue, eventValue) {
        if (timeValue && "object" == typeof timeValue) return this.dataToPosition(timeValue);
        const mockDatum = {};
        return this._timeField && (mockDatum[this._timeField] = timeValue), this._eventField && void 0 !== eventValue && (mockDatum[this._eventField] = eventValue), 
        this._getPoint(mockDatum);
    }
    getActiveMarks() {
        return [ this._axisLineMark, this._dotMark, this._iconMark, this._arrowMark, this._titleMark, this._subTitleMark ].filter(Boolean);
    }
}

exports.EventSeries = EventSeries, EventSeries.type = constant_1.EVENT_SERIES_TYPE, 
EventSeries.mark = eventSeriesMark, EventSeries.builtInTheme = {
    event: theme_1.event
}, EventSeries.transformerConstructor = vchart_1.BaseSeriesSpecTransformer;

const registerEventSeries = () => {
    (0, vchart_1.registerSymbolMark)(), (0, vchart_1.registerTextMark)(), (0, vchart_1.registerLineMark)(), 
    (0, vchart_1.registerPathMark)(), vchart_1.Factory.registerSeries(EventSeries.type, EventSeries);
};

exports.registerEventSeries = registerEventSeries;
