// @visactor/vchart-extension/cjs/charts/combination-candlestick/combination-candlestick-transformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.CombinationCandlestickChartSpecTransformer = void 0;

const vchart_1 = require("@visactor/vchart"), vchart_2 = require("@visactor/vchart"), candlestick_1 = require("../candlestick");

class CombinationCandlestickChartSpecTransformer extends vchart_1.CartesianChartSpecTransformer {
    needAxes() {
        return !0;
    }
    transformSpec(spec) {
        this.transformRegion(spec), void 0 === spec.tooltip && (spec.tooltip = {}), this.transformSeriesSpec(spec), 
        this._transformAxisSpec(spec), this._transformLayout(spec);
    }
    transformRegion(spec) {
        var _a, _b;
        const commonSpec = spec;
        commonSpec.region = [ null !== (_a = spec.candlestickRegion) && void 0 !== _a ? _a : {} ], 
        spec.previewSeries && commonSpec.region.push(null !== (_b = spec.previewRegion) && void 0 !== _b ? _b : {});
    }
    transformSeriesSpec(spec) {
        var _a;
        spec.series = null !== (_a = spec.series) && void 0 !== _a ? _a : [], spec.series.forEach((s => {
            this._transformSeriesData(spec, s), s.regionIndex = 0;
        }));
        const tempCandlestickSeries = spec.series.find((s => s.id === spec.candlestickSeries.id));
        let candlestickSpec;
        tempCandlestickSeries ? ((0, vchart_2.merge)(tempCandlestickSeries, spec.candlestickSeries), 
        candlestickSpec = tempCandlestickSeries) : (spec.series.push(spec.candlestickSeries), 
        candlestickSpec = spec.candlestickSeries), (0, candlestick_1.transformCandlestickSeriesSpec)(candlestickSpec), 
        this._transformSeriesData(spec, candlestickSpec), candlestickSpec.regionIndex = 0, 
        spec.previewSeries && (0, vchart_2.array)(spec.previewSeries).forEach((previewSeries => {
            const tempPreviewSeries = spec.series.find((s => s.id === previewSeries.id));
            let previewSeriesSpec;
            tempPreviewSeries ? ((0, vchart_2.merge)(tempPreviewSeries, previewSeries), previewSeriesSpec = tempPreviewSeries) : (spec.series.push(previewSeries), 
            previewSeriesSpec = previewSeries), previewSeriesSpec.regionIndex = 1, this._transformSeriesData(spec, previewSeriesSpec);
        }));
    }
    _transformSeriesData(spec, seriesSpec) {
        var _a;
        seriesSpec.data && (spec.data = (0, vchart_2.array)(null !== (_a = spec.data) && void 0 !== _a ? _a : []), 
        (0, vchart_2.array)(seriesSpec.data).forEach(((d, index) => {
            spec.data.push(d), 0 === index && (seriesSpec.dataIndex = spec.data.length - 1, 
            "id" in d ? seriesSpec.dataId = d.id : "name" in d && (seriesSpec.dataId = d.name));
        })), delete seriesSpec.data);
    }
    _transformAxisSpec(spec) {
        super._transformAxisSpec(spec), spec.previewSeries && (spec.previewAxes || (spec.previewAxes = {
            type: "linear",
            orient: "left",
            visible: !1
        }), spec.previewAxes.regionIndex = [ 1 ], spec.axes.push(spec.previewAxes), spec.axes.forEach((axis => {
            axis !== spec.previewAxes && ("left" !== axis.orient && "right" !== axis.orient || (axis.regionIndex = [ 0 ]));
        })));
    }
    _transformDataZoomSpec(spec) {
        spec.dataZoom && (0, vchart_2.array)(spec.dataZoom).forEach((dataZoom => {
            "left" !== dataZoom.orient && "right" !== dataZoom.orient || (dataZoom.regionIndex = [ 0 ]);
        }));
    }
    _transformLayout(spec) {
        var _a, _b;
        const layout = {
            type: "grid",
            col: 2,
            row: 6,
            elements: [],
            colWidth: [],
            rowHeight: []
        };
        let totalRow = 0, totalCol = 0;
        if (spec.title && !1 !== (null === (_a = spec.title) || void 0 === _a ? void 0 : _a.visible) && totalRow++, 
        spec.legends) {
            const firstLegend = (0, vchart_2.array)(spec.legends)[0];
            !1 !== (null == firstLegend ? void 0 : firstLegend.visible) && ("top" === firstLegend.orient || "bottom" === firstLegend.orient ? totalRow++ : totalCol++);
        }
        if (spec.dataZoom && (0, vchart_2.array)(spec.dataZoom).forEach((dataZoom => {
            !1 !== (null == dataZoom ? void 0 : dataZoom.visible) && ("top" === dataZoom.orient || "bottom" === dataZoom.orient ? totalRow++ : totalCol++);
        })), spec.axes) {
            const hasLayout = {
                top: !1,
                bottom: !1,
                left: !1,
                right: !1,
                z: !1
            };
            (0, vchart_2.array)(spec.axes).forEach((axis => {
                hasLayout[axis.orient] || (hasLayout[axis.orient] = !0, "top" === axis.orient || "bottom" === axis.orient ? totalRow++ : totalCol++);
            }));
        }
        totalRow++, totalCol++, spec.previewSeries && totalRow++, layout.row = totalRow, 
        layout.col = totalCol;
        const temp = {
            top: 0,
            bottom: 0,
            left: 0,
            right: 0
        };
        if (spec.title && !1 !== (null === (_b = spec.title) || void 0 === _b ? void 0 : _b.visible) && this._layoutOrientItem(layout, spec.title, temp, {
            specKey: "title",
            index: 0
        }, "total"), spec.legends) {
            const firstLegend = (0, vchart_2.array)(spec.legends)[0];
            !1 !== (null == firstLegend ? void 0 : firstLegend.visible) && this._layoutOrientItem(layout, firstLegend, temp, {
                specKey: "legends",
                index: 0
            }, "less");
        }
        if (spec.dataZoom && (0, vchart_2.array)(spec.dataZoom).forEach(((dataZoom, index) => {
            !1 !== (null == dataZoom ? void 0 : dataZoom.visible) && this._layoutOrientItem(layout, dataZoom, temp, {
                specKey: "dataZoom",
                index: index
            }, "one");
        })), spec.axes) {
            const hasLayout = {
                left: !1,
                right: !1
            };
            spec.axes.forEach(((axis, index) => {
                "left" !== axis.orient && "right" !== axis.orient || hasLayout[axis.orient] || (hasLayout[axis.orient] = !0, 
                this._layoutOrientItem(layout, axis, temp, {
                    specKey: "axes",
                    index: index
                }, "one"));
            }));
        }
        const regionLayout = this._layoutOrientItem(layout, {
            orient: "top"
        }, temp, {
            specKey: "region",
            index: 0
        }, "one");
        if (spec.previewSeries) {
            const regionLayout = this._layoutOrientItem(layout, {
                orient: "bottom"
            }, temp, {
                specKey: "region",
                index: 1
            }, "one"), previewAxesIndex = spec.axes.findIndex((axis => axis === spec.previewAxes));
            if (-1 !== previewAxesIndex) {
                const previewAxesLayout = {
                    col: regionLayout.col - 1,
                    row: regionLayout.row,
                    modelKey: "axes",
                    modelIndex: previewAxesIndex
                };
                layout.elements.push(previewAxesLayout);
            }
            if (spec.previewHeight) if ("string" == typeof spec.previewHeight) {
                if ((0, vchart_1.isPercent)(spec.previewHeight)) {
                    const heightStr = spec.previewHeight;
                    layout.rowHeight.push({
                        index: regionLayout.row,
                        size: size => Number(heightStr.substring(0, heightStr.length - 1)) * size / 100
                    });
                }
            } else layout.rowHeight.push({
                index: regionLayout.row,
                size: spec.previewHeight
            });
        }
        const bottomAxesIndex = spec.axes.findIndex((axis => "bottom" === axis.orient));
        this._layoutOrientItem(layout, spec.axes[bottomAxesIndex], temp, {
            specKey: "axes",
            index: bottomAxesIndex
        }, "one"), spec.axes.forEach(((axis, index) => {
            if (axis === spec.previewAxes) return;
            const axisLayout = layout.elements.find((item => "modelKey" in item && "axes" === item.modelKey && item.modelIndex === index));
            axisLayout && ("top" === axis.orient || "bottom" === axis.orient ? axisLayout.col = regionLayout.col : axisLayout.row = regionLayout.row);
        })), (0, vchart_2.array)(spec.dataZoom).forEach(((dataZoom, index) => {
            const dataZoomLayout = layout.elements.find((item => "modelKey" in item && "dataZoom" === item.modelKey && item.modelIndex === index));
            dataZoomLayout && ("top" === dataZoom.orient || "bottom" === dataZoom.orient ? dataZoomLayout.col = regionLayout.col : dataZoomLayout.row = regionLayout.row);
        })), spec.layout = layout;
    }
    _layoutOrientItem(layout, modelSpec, temp, modelInfo, spanType) {
        var _a;
        const modelLayout = {
            modelKey: modelInfo.specKey,
            modelIndex: modelInfo.index,
            col: 0,
            row: 0
        }, orient = null !== (_a = modelSpec.orient) && void 0 !== _a ? _a : "top", {span: span, index: index} = this._getLayoutElementCommon(layout, temp, spanType, orient);
        return layout.elements.push(modelLayout), "top" === orient ? (modelLayout.row = temp.top, 
        modelLayout.rowSpan = 1, modelLayout.colSpan = span, modelLayout.col = index, temp.top++) : "bottom" === orient ? (modelLayout.row = layout.row - temp.bottom - 1, 
        modelLayout.rowSpan = 1, modelLayout.colSpan = span, modelLayout.col = index, temp.bottom++) : "left" === orient ? (modelLayout.col = temp.left, 
        modelLayout.colSpan = 1, modelLayout.rowSpan = span, modelLayout.row = index, temp.left++) : "right" === orient && (modelLayout.col = layout.col - temp.right - 1, 
        modelLayout.colSpan = 1, modelLayout.rowSpan = span, modelLayout.row = index, temp.right++), 
        modelLayout;
    }
    _getLayoutElementCommon(layout, temp, span, orient) {
        return "top" === orient || "bottom" === orient ? {
            span: "total" === span ? layout.col : "one" === span ? 1 : layout.col - temp.left - temp.right,
            index: "total" === span ? 0 : temp.left
        } : {
            span: "total" === span ? layout.row : "one" === span ? 1 : layout.row - temp.top - temp.bottom,
            index: "total" === span ? 0 : temp.top
        };
    }
}

exports.CombinationCandlestickChartSpecTransformer = CombinationCandlestickChartSpecTransformer;
